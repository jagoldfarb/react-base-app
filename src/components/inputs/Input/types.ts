export interface IInputProps {
  disabled?: boolean;
  input: {
    onChange: (value: string) => void;
    value: string;
  };
  label?: string;
  maxLength?: number;
  meta: {
    touched: boolean;
    error?: string;
  };
  rows?: number;
  rowsMax?: number;
  size?: 'medium' | 'small' | undefined;
  type?: string;
  variant?: 'outlined' | 'standard' | 'filled' | undefined;
}
