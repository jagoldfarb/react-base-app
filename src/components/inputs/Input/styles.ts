import styled from 'styled-components';
import { TextField } from '@material-ui/core';

import { COLORS } from 'helpers/enums/colors';
import { SIZES } from 'helpers/enums/sizes';

export const FormGroup = styled.div`
  width: 100%;
  position: relative;
  padding-bottom: 35px;

  @media (max-width: ${SIZES.mediumPhone}px) {
    padding-bottom: 45px;
  }
`;

export const CustomInput = styled(TextField)`
  width: 100%;
  height: fit-content;
  margin-bottom: 0px !important;

  label {
    display: flex;
    font-size: 14px;
    span {
      display: none;
    }
  }

  .MuiInputBase-input {
    font-size: 14px;

    &:-webkit-autofill,
    &:-webkit-autofill:hover,
    &:-webkit-autofill::first-line,
    &:-webkit-autofill:focus {
      -webkit-box-shadow: 0 0 0px 1000px white inset !important;
      background-color: white;
      background-clip: content-box !important;
      font-size: 14px;
    }
  }
`;

export const ErrorBlock = styled.div`
  width: 100%;
  position: absolute;
  font-size: 13px;
  line-height: 1.5;
  color: ${COLORS.red};
  text-align: left;

  @media (max-width: ${SIZES.mediumPhone}px) {
    font-size: 12px;
  }
`;
