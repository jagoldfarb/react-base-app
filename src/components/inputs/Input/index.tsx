import React, { useState, ChangeEvent, FunctionComponent } from 'react';
import { InputAdornment, IconButton } from '@material-ui/core';
import { Visibility, VisibilityOff } from '@material-ui/icons';

import { IInputProps } from './types';
import { FormGroup, CustomInput, ErrorBlock } from './styles';

const Input: FunctionComponent<IInputProps> = (props: IInputProps) => {
  const {
    disabled = false,
    input,
    label = '',
    maxLength = 255,
    meta: { touched, error },
    rows = 3,
    rowsMax = 3,
    size = 'medium',
    type = 'text',
    variant = 'outlined',
  } = props;
  const [showPassword, setShowPassword] = useState<boolean>(false);
  const { onChange, value } = input;
  const isPasswordInput = type === 'password';
  const isTextAreaInput = type === 'textarea';
  const hasError = touched && !!error;

  const getInputType = () => {
    let inputType = type;
    if (isPasswordInput) {
      inputType = showPassword ? 'text' : type;
    }
    return inputType;
  };

  return (
    <FormGroup>
      <CustomInput
        {...input}
        disabled={disabled}
        error={hasError}
        inputProps={{ 'data-testid': `input-${label}`, maxLength }}
        label={label}
        multiline={isTextAreaInput}
        onChange={(event: ChangeEvent<HTMLInputElement>) => onChange(event.target.value)}
        rows={rows}
        rowsMax={rowsMax}
        size={size}
        type={getInputType()}
        value={value}
        variant={variant}
        // eslint-disable-next-line react/jsx-no-duplicate-props
        InputProps={{
          endAdornment: isPasswordInput && (
            <InputAdornment position="end">
              <IconButton
                aria-label="toggle password visibility"
                data-testid="toggle-password-button"
                onClick={() => setShowPassword(!showPassword)}
              >
                {showPassword ? <VisibilityOff /> : <Visibility />}
              </IconButton>
            </InputAdornment>
          ),
        }}
      />
      {hasError && <ErrorBlock data-testid="error-block">{error}</ErrorBlock>}
    </FormGroup>
  );
};

export default Input;
