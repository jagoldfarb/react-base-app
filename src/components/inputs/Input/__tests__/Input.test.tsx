import React from 'react';
import { render, fireEvent } from '@testing-library/react';

import Input from '..';

const defaultProps = {
  input: {
    name: 'email',
    onChange: jest.fn(),
    value: '',
  },
  label: 'Email',
  meta: {
    error: '',
    touched: false,
  },
};

describe('Input Component', () => {
  it('Should render the input', () => {
    const { asFragment, getByTestId } = render(<Input {...defaultProps} />);
    expect(getByTestId('input-Email')).toBeDefined();
    expect(asFragment()).toMatchSnapshot();
  });

  it('Should render the input with no label', () => {
    const props = { ...defaultProps, label: undefined };
    const { asFragment, getByTestId } = render(<Input {...props} />);
    expect(getByTestId('input-')).toBeDefined();
    expect(asFragment()).toMatchSnapshot();
  });

  it('Should render the input as a password input', () => {
    const props = {
      ...defaultProps,
      label: 'Password',
      type: 'password',
    };
    const { asFragment, getByTestId } = render(<Input {...props} />);
    expect(getByTestId('input-Password')).toBeDefined();
    // Press toggle password button
    const togglePasswordInput = getByTestId('toggle-password-button');
    expect(togglePasswordInput).toBeDefined();
    fireEvent.click(togglePasswordInput);
    expect(asFragment()).toMatchSnapshot();
  });

  it('Should render the input as a textarea input', () => {
    const props = {
      ...defaultProps,
      label: 'Message',
      type: 'textarea',
    };
    const { asFragment, getByTestId } = render(<Input {...props} />);
    expect(getByTestId('input-Message')).toBeDefined();
    expect(asFragment()).toMatchSnapshot();
  });

  it('Should render the input and change the value', () => {
    const { asFragment, getByTestId } = render(<Input {...defaultProps} />);
    const input = getByTestId('input-Email');
    expect(input).toBeDefined();
    fireEvent.change(input, { target: { value: 'info@gmail.com' } });
    expect(asFragment()).toMatchSnapshot();
  });

  it('Should render the input with errors', () => {
    const props = {
      ...defaultProps,
      meta: {
        error: 'Required',
        touched: true,
      },
    };
    const { asFragment, getByTestId, getByText } = render(<Input {...props} />);
    expect(getByTestId('input-Email')).toBeDefined();
    expect(getByTestId('error-block')).toBeDefined();
    expect(getByText('Required')).toBeDefined();
    expect(asFragment()).toMatchSnapshot();
  });
});
