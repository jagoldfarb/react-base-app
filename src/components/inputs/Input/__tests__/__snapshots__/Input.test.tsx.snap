// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input Component Should render the input 1`] = `
<DocumentFragment>
  <div
    class="sc-bdfBwQ cgTScR"
  >
    <div
      class="MuiFormControl-root MuiTextField-root sc-gsTCUz jKtOdX"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
        data-shrink="false"
      >
        Email
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input"
          data-testid="input-Email"
          maxlength="255"
          name="email"
          rows="3"
          type="text"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="PrivateNotchedOutline-legendLabelled-3"
          >
            <span>
              Email
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Input Component Should render the input and change the value 1`] = `
<DocumentFragment>
  <div
    class="sc-bdfBwQ cgTScR"
  >
    <div
      class="MuiFormControl-root MuiTextField-root sc-gsTCUz jKtOdX"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
        data-shrink="false"
      >
        Email
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input"
          data-testid="input-Email"
          maxlength="255"
          name="email"
          rows="3"
          type="text"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="PrivateNotchedOutline-root-17 MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="PrivateNotchedOutline-legendLabelled-19"
          >
            <span>
              Email
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Input Component Should render the input as a password input 1`] = `
<DocumentFragment>
  <div
    class="sc-bdfBwQ cgTScR"
  >
    <div
      class="MuiFormControl-root MuiTextField-root sc-gsTCUz jKtOdX"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
        data-shrink="false"
      >
        Password
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl MuiInputBase-adornedEnd MuiOutlinedInput-adornedEnd"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd MuiOutlinedInput-inputAdornedEnd"
          data-testid="input-Password"
          maxlength="255"
          name="email"
          rows="3"
          type="text"
          value=""
        />
        <div
          class="MuiInputAdornment-root MuiInputAdornment-positionEnd"
        >
          <button
            aria-label="toggle password visibility"
            class="MuiButtonBase-root MuiIconButton-root"
            data-testid="toggle-password-button"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 7c2.76 0 5 2.24 5 5 0 .65-.13 1.26-.36 1.83l2.92 2.92c1.51-1.26 2.7-2.89 3.43-4.75-1.73-4.39-6-7.5-11-7.5-1.4 0-2.74.25-3.98.7l2.16 2.16C10.74 7.13 11.35 7 12 7zM2 4.27l2.28 2.28.46.46C3.08 8.3 1.78 10.02 1 12c1.73 4.39 6 7.5 11 7.5 1.55 0 3.03-.3 4.38-.84l.42.42L19.73 22 21 20.73 3.27 3 2 4.27zM7.53 9.8l1.55 1.55c-.05.21-.08.43-.08.65 0 1.66 1.34 3 3 3 .22 0 .44-.03.65-.08l1.55 1.55c-.67.33-1.41.53-2.2.53-2.76 0-5-2.24-5-5 0-.79.2-1.53.53-2.2zm4.31-.78l3.15 3.15.02-.16c0-1.66-1.34-3-3-3l-.17.01z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
        <fieldset
          aria-hidden="true"
          class="PrivateNotchedOutline-root-9 MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="PrivateNotchedOutline-legendLabelled-11"
          >
            <span>
              Password
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Input Component Should render the input as a textarea input 1`] = `
<DocumentFragment>
  <div
    class="sc-bdfBwQ cgTScR"
  >
    <div
      class="MuiFormControl-root MuiTextField-root sc-gsTCUz jKtOdX"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
        data-shrink="false"
      >
        Message
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl MuiInputBase-multiline MuiOutlinedInput-multiline"
      >
        <textarea
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline MuiOutlinedInput-inputMultiline"
          data-testid="input-Message"
          maxlength="255"
          name="email"
          rows="3"
          style="height: 0px; overflow: hidden;"
        />
        <textarea
          aria-hidden="true"
          class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline MuiOutlinedInput-inputMultiline"
          readonly=""
          style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); width: 100%;"
          tabindex="-1"
        />
        <fieldset
          aria-hidden="true"
          class="PrivateNotchedOutline-root-13 MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="PrivateNotchedOutline-legendLabelled-15"
          >
            <span>
              Message
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Input Component Should render the input with errors 1`] = `
<DocumentFragment>
  <div
    class="sc-bdfBwQ cgTScR"
  >
    <div
      class="MuiFormControl-root MuiTextField-root sc-gsTCUz jKtOdX"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-error Mui-error"
        data-shrink="false"
      >
        Email
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root Mui-error Mui-error MuiInputBase-formControl"
      >
        <input
          aria-invalid="true"
          class="MuiInputBase-input MuiOutlinedInput-input"
          data-testid="input-Email"
          maxlength="255"
          name="email"
          rows="3"
          type="text"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="PrivateNotchedOutline-root-21 MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="PrivateNotchedOutline-legendLabelled-23"
          >
            <span>
              Email
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      class="sc-dlfnbm bOmIgN"
      data-testid="error-block"
    >
      Required
    </div>
  </div>
</DocumentFragment>
`;

exports[`Input Component Should render the input with no label 1`] = `
<DocumentFragment>
  <div
    class="sc-bdfBwQ cgTScR"
  >
    <div
      class="MuiFormControl-root MuiTextField-root sc-gsTCUz jKtOdX"
    >
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input"
          data-testid="input-"
          maxlength="255"
          name="email"
          rows="3"
          type="text"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="PrivateNotchedOutline-root-5 MuiOutlinedInput-notchedOutline"
          style="padding-left: 8px;"
        >
          <legend
            class="PrivateNotchedOutline-legend-6"
            style="width: 0.01px;"
          >
            <span>
              â€‹
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
